Rewrite MemoryUsage to not use recursion

fingolfin opened the pull request:
This was motivated by PR #1522, where `MemoryUsage` failed on a deeply nested recursive data structure. With this PR, it works on the example in that PR.

ChrisJefferson commented:
My only comment would be to mention in the commit you've also substantially changed the algorithm in several ways. I agree it's better, but it's not just "removing the recursion", which would have been a much smaller change.

fingolfin commented:
You are right, will do!

fingolfin commented:
So, something roughly like 'replace recursion by a work queue, which enables dealing with deeply nested inputs. To still make it extensible, add MemoryUsageOp ... yadda yadda' ?

ChrisJefferson commented:
Yes, just to make it clear there is more here.

At first I was confused the logic has changed, as we used to skip things in lists with SHALLOW_SIZE=0, but now that's moved into visit, where we just skip T_INT and T_FFE. Which is the same thing of course, but it took me a minute to convince myself of that. The reasonably extensive tests hopefully demonstrate all this hasn't changed any logic anyway.

No need to put that in the comment, just that there is more here :)

fingolfin commented:
Actually, there is a change, because non-immediate objects with shallow size 0 can exist, and we now handle them correctly (that was an explicit decision, too)

fingolfin commented:
Updated. I also changed `MemoryUsageOp` to return a value which is added to the total, to allow packages with external memory use to also add that in.
